### ISSUE TITLE
Proposal: merge of resources from multiple config files

### ISSUE THREAD

**Issue by inercia** [hashicorp/terraform#15190]
### Current situation

When defining a resource like this:

```hcl
# cluster.tf
resource "libvirt_domain" "admin_node" {
  name       = "admin"
  memory     = "2048"

  network_interface {
    network_name   = "default"
  }

  #...
}
```

we cannot add more behaviour to `admin_node` by redefining the same resource in another file and adding, for example, a provisioner this way:

```hcl
# cluster.local.tf
resource "libvirt_domain" "admin_node" {
  provisioner "remote-exec" {
    inline = [
      "echo 'we could do some local changes'",
    ]
  }
}
```
Both files will be merged by Terraform, but it fails as it detects that `admin_resource` has been defined multiple times.

```
$ terraform plan
1 error(s) occurred:

* module root: 1 error(s) occurred:

* libvirt_domain.admin: resource repeated multiple times
```

### Proposal

Terraform could ignore that both files define the same resource and try to do a _deep merge_, merging all the resources found in the files. This could allow users to add additional behaviour to an already defined Terraform file.

### Implementation

I could investigate how to implement this stuff...

**Comment by apparentlymart** [comment]
Hi @inercia! Thanks for this proposal.

This is an interesting idea. A concern I have with it is that it becomes hard to see easily all the settings for a particular resource. Rather than locating its heading in one file and reading it there, one must instead consult _all_ files.

Furthermore, we have plans to improve the error messages Terraform produces to specify the location of a resource that has an error, but that is possible only if each resource has only a single definition point.

I'd be interested to hear more about your use-case (the problem you have encountered) as opposed to this specific solution to the problem; there may be other ways to address the problem that preserve the above goals.

**Comment by inercia** [comment]
In my particular use case, we have a _base_ terraform file, `cluster.tf`, that is used for creating our cluster, but depending on the member of the team they would like to add some extra behaviour:

* regular users just use `cluster.tf` for creating the cluster
* some developers want to test their `feature_x`, so they would like to copy their own code and scripts to the very same machines created by `cluster.tf`, so they would like to have have `john-feature-x.local.tf` (not in the repo).
* the qa guys must also provision some extra stuff to this machines and run their stuff. In some cases, they also want to test particular features in a different file, `test-feature-x.local.tf`.

So far the only solution has been to play with the `count` parameter for `resources` and duplicate the whole resource content (or try to move stuff to modules), but that makes the whole `.tf` a mess and makes it impossible to add unforeseen `local.tf` files.

I don't think defining the same resource in multiple files would make things more difficult to understand, but that could be solved if plan could show some information about the resource being defined as a result of a merge. I understand that maybe it would make things a bit more difficult for error reporting, but I think the benefits would pay off...

**Comment by apparentlymart** [comment]
Thanks for sharing these details @inercia!

Is it correct to assume that in the situation you're describing it's specifically `provisioner` blocks that need to be added to these resources, to deal with configuring the created machines?

If so, we do currently have a (admittedly non-intuitive) workaround for such use-cases, in the form of the `null_resource` resource that allows provisioners to be run separately from what they are provisioning:

```hcl
resource "libvirt_domain" "admin_node" {
  name       = "admin"
  memory     = "2048"

  network_interface {
    network_name   = "default"
  }

  #...
}

resource "null_resource" "provision" {
  connection {
    # (connection information to access the libvirt_domain)
  }
  provisioner "remote-exec" {
    inline = [
      "echo 'we could do some local changes'",
    ]
  }
}
```

We know this is not the most ideal way to express this, and have discussed before the idea of allowing provisioners _specifically_ to be declared at top-level, but currently the null resource is used as a solution to the problem of defining exactly what the lifecycle for such a top-level provisioner should be; by wrapping 
it in a do-nothing resource we can re-use the resource lifecycle, allowing e.g. the use of `taint` to force the provisioner to re-run, etc.

**Comment by sjwl** [comment]
could [overrides](https://www.terraform.io/docs/configuration/override.html) help?  Each user could bring in their own override to customize the specific resource.  If I understand the terraform logic, overrides conduct a _merge_ rather than an _append_ as to avoid the duplicate resource?

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

I think @sjwl is right that [override files](https://www.terraform.io/docs/configuration/override.html) are the solution to this use-case. There are some quirks with using them on certain constructs in the v0.11 and prior releases but the forthcoming v0.12.0 release includes a more robust implementation of override files that should behave in an intuitive way for all constructs. Since that change is already merged in master ready for the later release, I'm going to close this out now.

There is also [some extended documentation for Override Files](https://github.com/hashicorp/terraform/blob/a63c9d7db5ecebdaa79dbfc4ed29606fd2b573d2/website/docs/configuration/override.html.md) ready to be published as part of the final v0.12.0 release.

Thanks for suggesting this, and sorry we missed updating this when the override files feature was introduced.
